File: src/modules/trigger_engine.rs
use super::triggers::{TriggerAction, TriggerEvent, TriggerRule, TriggersConfig};
use anyhow::{Context, Result};
use std::collections::HashMap;
use std::sync::Arc;
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use tokio::sync::Mutex;
use tracing::{info, warn, error};

/// Triggered event record
#[derive(Debug, Clone)]
pub struct TriggeredEvent {
    pub trigger_name: String,
    pub timestamp: u64,
    pub message: String,
    pub action: TriggerAction,
}

/// Trigger engine with rate limiting
pub struct TriggerEngine {
    config: TriggersConfig,
    last_triggered: Arc<Mutex<HashMap<String, u64>>>, // trigger_name -> last_timestamp
    triggered_events: Arc<Mutex<Vec<TriggeredEvent>>>,
}

impl TriggerEngine {
    /// Create new trigger engine
    pub fn new(config: TriggersConfig) -> Self {
        Self {
            config,
            last_triggered: Arc::new(Mutex::new(HashMap::new())),
            triggered_events: Arc::new(Mutex::new(Vec::new())),
        }
    }

    /// Reload configuration
    pub async fn reload_config(&mut self, config: TriggersConfig) {
        info!("Reloading trigger configuration ({} rules)", config.triggers.len());
        self.config = config;
    }

    /// Evaluate all triggers against an event
    pub async fn evaluate(&self, event: TriggerEvent) -> Result<Vec<TriggeredEvent>> {
        let mut triggered = Vec::new();
        let now = current_timestamp();

        for (name, rule) in &self.config.triggers {
            // Check if trigger matches
            if !rule.matches(&event) {
                continue;
            }

            // Check rate limiting
            let mut last_triggered = self.last_triggered.lock().await;
            if let Some(&last_time) = last_triggered.get(name) {
                let elapsed = now.saturating_sub(last_time);
                if elapsed < rule.cooldown_seconds {
                    continue; // Still in cooldown
                }
            }

            // Trigger matched and not in cooldown
            let message = format_message(rule, &event);

            info!("🔔 Trigger '{}' activated: {}", name, message);

            // Execute action
            match self.execute_action(name, rule, &message).await {
                Ok(_) => {
                    // Update last triggered time
                    last_triggered.insert(name.clone(), now);

                    // Record triggered event
                    let triggered_event = TriggeredEvent {
                        trigger_name: name.clone(),
                        timestamp: now,
                        message: message.clone(),
                        action: rule.action.clone(),
                    };

                    triggered.push(triggered_event.clone());

                    // Store in history
                    let mut history = self.triggered_events.lock().await;
                    history.push(triggered_event);

                    // Keep only last 1000 events
                    if history.len() > 1000 {
                        history.remove(0);
                    }
                }
                Err(e) => {
                    error!("Failed to execute trigger '{}': {}", name, e);
                }
            }
        }

        Ok(triggered)
    }

    /// Execute trigger action
    async fn execute_action(&self, name: &str, rule: &TriggerRule, message: &str) -> Result<()> {
        match rule.action {
            TriggerAction::Notify => {
                self.send_notification(name, message).await?;
            }
            TriggerAction::Log => {
                self.log_trigger(name, message).await?;
            }
            TriggerAction::Command => {
                if let Some(cmd) = &rule.command {
                    self.execute_command(name, cmd, message).await?;
                } else {
                    warn!("Trigger '{}' has 'command' action but no command specified", name);
                }
            }
        }
        Ok(())
    }

    /// Send desktop notification
    async fn send_notification(&self, name: &str, message: &str) -> Result<()> {
        #[cfg(target_os = "linux")]
        {
            // Use notify-send on Linux
            let output = tokio::process::Command::new("notify-send")
                .arg("Raven Trigger")
                .arg(format!("{}: {}", name, message))
                .arg("--urgency=normal")
                .arg("--icon=dialog-information")
                .output()
                .await;

            match output {
                Ok(result) => {
                    if !result.status.success() {
                        warn!("notify-send failed: {}", String::from_utf8_lossy(&result.stderr));
                    }
                }
                Err(e) => {
                    warn!("Failed to execute notify-send: {}", e);
                }
            }
        }

        #[cfg(target_os = "macos")]
        {
            // Use osascript on macOS
            let script = format!(
                "display notification \"{}\" with title \"Raven Trigger: {}\"",
                message, name
            );
            let output = tokio::process::Command::new("osascript")
                .arg("-e")
                .arg(script)
                .output()
                .await;

            match output {
                Ok(result) => {
                    if !result.status.success() {
                        warn!("osascript failed: {}", String::from_utf8_lossy(&result.stderr));
                    }
                }
                Err(e) => {
                    warn!("Failed to execute osascript: {}", e);
                }
            }
        }

        #[cfg(target_os = "windows")]
        {
            // Use PowerShell on Windows
            let ps_command = format!(
                "[Windows.UI.Notifications.ToastNotificationManager, Windows.UI.Notifications, ContentType = WindowsRuntime] | Out-Null; \
                $template = [Windows.UI.Notifications.ToastNotificationManager]::GetTemplateContent([Windows.UI.Notifications.ToastTemplateType]::ToastText02); \
                $template.SelectSingleNode('//text[@id=\"1\"]').InnerText = 'Raven Trigger: {}'; \
                $template.SelectSingleNode('//text[@id=\"2\"]').InnerText = '{}'; \
                $toast = [Windows.UI.Notifications.ToastNotification]::new($template); \
                [Windows.UI.Notifications.ToastNotificationManager]::CreateToastNotifier('Raven').Show($toast);",
                name, message
            );

            let output = tokio::process::Command::new("powershell")
                .arg("-Command")
                .arg(ps_command)
                .output()
                .await;

            match output {
                Ok(result) => {
                    if !result.status.success() {
                        warn!("PowerShell notification failed: {}", String::from_utf8_lossy(&result.stderr));
                    }
                }
                Err(e) => {
                    warn!("Failed to execute PowerShell: {}", e);
                }
            }
        }

        info!("📢 Notification sent: {} - {}", name, message);
        Ok(())
    }

    /// Log trigger to file
    async fn log_trigger(&self, name: &str, message: &str) -> Result<()> {
        use tokio::io::AsyncWriteExt;

        let log_path = std::env::current_dir()
            .context("Failed to get current directory")?
            .join(".raven")
            .join("triggers.log");

        let timestamp = chrono::Utc::now().to_rfc3339();
        let log_line = format!("[{}] {} - {}\n", timestamp, name, message);

        let mut file = tokio::fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&log_path)
            .await
            .context("Failed to open triggers.log")?;

        file.write_all(log_line.as_bytes())
            .await
            .context("Failed to write to triggers.log")?;

        info!("📝 Logged trigger: {} - {}", name, message);
        Ok(())
    }

    /// Execute shell command
    async fn execute_command(&self, name: &str, command: &str, message: &str) -> Result<()> {
        info!("⚙️ Executing command for trigger '{}': {}", name, command);

        let output = tokio::process::Command::new("sh")
            .arg("-c")
            .arg(command)
            .output()
            .await
            .context("Failed to execute command")?;

        if output.status.success() {
            info!("✅ Command succeeded for trigger '{}'", name);
        } else {
            warn!(
                "❌ Command failed for trigger '{}': {}",
                name,
                String::from_utf8_lossy(&output.stderr)
            );
        }

        Ok(())
    }

    /// Get recent triggered events
    pub async fn get_recent_triggers(&self, limit: usize) -> Vec<TriggeredEvent> {
        let history = self.triggered_events.lock().await;
        let start = history.len().saturating_sub(limit);
        history[start..].to_vec()
    }

    /// Get trigger statistics
    pub async fn get_stats(&self) -> TriggerStats {
        let history = self.triggered_events.lock().await;
        let last_triggered = self.last_triggered.lock().await;

        let mut trigger_counts: HashMap<String, usize> = HashMap::new();
        for event in history.iter() {
            *trigger_counts.entry(event.trigger_name.clone()).or_insert(0) += 1;
        }

        TriggerStats {
            total_triggers: history.len(),
            active_triggers: self.config.triggers.len(),
            trigger_counts,
            cooldowns: last_triggered.clone(),
        }
    }

    /// Clear all rate limiting state
    pub async fn clear_cooldowns(&self) {
        let mut last_triggered = self.last_triggered.lock().await;
        last_triggered.clear();
        info!("Cleared all trigger cooldowns");
    }
}

/// Trigger statistics
#[derive(Debug, Clone)]
pub struct TriggerStats {
    pub total_triggers: usize,
    pub active_triggers: usize,
    pub trigger_counts: HashMap<String, usize>,
    pub cooldowns: HashMap<String, u64>,
}

/// Format message with event placeholders
fn format_message(rule: &TriggerRule, event: &TriggerEvent) -> String {
    let base_message = rule.message.as_deref().unwrap_or("Trigger activated");

    let mut message = base_message.to_string();

    // Replace placeholders
    if let Some(file) = &event.file {
        message = message.replace("{file}", file);
    }
    if let Some(agent) = &event.agent {
        message = message.replace("{agent}", agent);
    }
    if let Some(event_type) = &event.event_type {
        message = message.replace("{event_type}", event_type);
    }
    if let Some(lines) = event.lines_changed {
        message = message.replace("{lines_changed}", &lines.to_string());
    }
    if let Some(duration) = event.duration_ms {
        message = message.replace("{duration_ms}", &duration.to_string());
    }
    if let Some(cpu) = event.cpu_percent {
        message = message.replace("{cpu_percent}", &format!("{:.1}", cpu));
    }
    if let Some(mem) = event.memory_percent {
        message = message.replace("{memory_percent}", &format!("{:.1}", mem));
    }

    message
}

/// Get current Unix timestamp
fn current_timestamp() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_secs()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::modules::triggers::TriggerAction;

    #[tokio::test]
    async fn test_trigger_engine_creation() {
        let config = TriggersConfig::default();
        let engine = TriggerEngine::new(config);
        let stats = engine.get_stats().await;
        assert_eq!(stats.active_triggers, 0);
    }

    #[tokio::test]
    async fn test_message_formatting() {
        let rule = TriggerRule {
            file: Some("*.rs".to_string()),
            agent: None,
            event_type: None,
            lines_changed: None,
            duration_ms: None,
            cpu_percent: None,
            memory_percent: None,
            action: TriggerAction::Notify,
            message: Some("File {file} changed with {lines_changed} lines".to_string()),
            command: None,
            cooldown_seconds: 60,
        };

        let event = TriggerEvent {
            file: Some("main.rs".to_string()),
            agent: None,
            event_type: None,
            lines_changed: Some(150),
            duration_ms: None,
            cpu_percent: None,
            memory_percent: None,
        };

        let message = format_message(&rule, &event);
        assert_eq!(message, "File main.rs changed with 150 lines");
    }
}

File: backend/trigger-engine.ts
/**
 * TriggerEngine - Custom alert trigger system
 *
 * Evaluates rules from config.toml and fires triggers based on:
 * - File patterns
 * - Agent names
 * - Event types
 * - Numeric thresholds (lines changed, duration, CPU, memory)
 *
 * Integrates with EventBus for event-driven architecture
 */

import fs from 'fs';
import path from 'path';
import { execFile } from 'child_process';
import { promisify } from 'util';
import { logger } from './utils/logger.js';
import toml from 'toml';
import { EventBus, TriggerFiredEvent } from './modules/eventBus.js';
import { internalMetrics } from './services/internal-metrics.js';
import type { Server as SocketIOServer } from 'socket.io';

const execFileAsync = promisify(execFile);

// ==================== Type Definitions ====================

/**
 * Trigger rule definition from config.toml
 */
interface TriggerRule {
  name: string;
  file?: string;
  agent?: string;
  event_type?: string;
  lines_changed?: string;
  lines_deleted?: string;
  duration_ms?: string;
  cpu_percent?: string;
  memory_percent?: string;
  action: 'log' | 'notify' | 'command';
  message?: string;
  command?: string;
  cooldown_seconds: number;
}

/**
 * Event data passed to trigger evaluation
 */
interface TriggerEvent {
  file?: string;
  agent?: string;
  event_type?: string;
  lines_changed?: number;
  duration_ms?: number;
  cpu_percent?: number;
  memory_percent?: number;
  event_size?: number;
}

/**
 * Triggered event record
 */
interface TriggeredEventRecord {
  trigger_name: string;
  timestamp: number;
  message: string;
  action: string;
}

/**
 * Trigger statistics
 */
interface TriggerStats {
  total_triggers: number;
  active_triggers: number;
  trigger_counts: Record<string, number>;
}

// ==================== TriggerEngine Class ====================

export class TriggerEngine {
  private configPath: string;
  private io: SocketIOServer | null;
  private triggers: Map<string, TriggerRule>;
  private triggeredEvents: TriggeredEventRecord[];
  private cooldowns: Map<string, number>;
  private triggerCounts: Map<string, number>;

  constructor(configPath: string, io: SocketIOServer | null = null) {
    this.configPath = configPath;
    this.io = io;
    this.triggers = new Map();
    this.triggeredEvents = [];
    this.cooldowns = new Map();
    this.triggerCounts = new Map();
    this.loadConfig();
  }

  /**
   * Set Socket.IO instance for real-time updates
   */
  setIo(io: SocketIOServer): void {
    this.io = io;
  }

  /**
   * Load triggers from config.toml
   */
  loadConfig(): void {
    try {
      const configFile = path.join(this.configPath, 'config.toml');

      if (!fs.existsSync(configFile)) {
        logger.info('⚠️  No config.toml found, creating example config...');
        this.createExampleConfig(configFile);
      }

      const configContent = fs.readFileSync(configFile, 'utf8');
      const config = toml.parse(configContent);

      this.triggers.clear();

      if (config.triggers) {
        for (const [name, rule] of Object.entries(config.triggers)) {
          this.triggers.set(name, {
            name,
            ...(rule as Omit<TriggerRule, 'name'>),
            cooldown_seconds: (rule as any).cooldown_seconds || 60
          });
        }
      }

      logger.info(`✅ Loaded ${this.triggers.size} trigger rules`);
    } catch (error: any) {
      logger.error('❌ Failed to load triggers config:', error);
    }
  }

  /**
   * Create example config file
   */
  private createExampleConfig(configFile: string): void {
    const exampleConfig = `# Raven Custom Triggers Configuration

[triggers.large_edit]
file = "*.js"
lines_changed = ">100"
action = "log"
message = "Large edit detected: {file} ({lines_changed} lines)"
cooldown_seconds = 60

[triggers.slow_operation]
agent = "claude"
duration_ms = ">5000"
action = "log"
message = "Slow {agent} operation: {file} took {duration_ms}ms"
cooldown_seconds = 120

[triggers.high_cpu]
cpu_percent = ">80"
action = "log"
message = "High CPU usage: {cpu_percent}%"
cooldown_seconds = 300
`;

    fs.mkdirSync(path.dirname(configFile), { recursive: true });
    fs.writeFileSync(configFile, exampleConfig);
    logger.info(`✅ Created example config at ${configFile}`);
  }

  /**
   * Evaluate an event against all trigger rules
   */
  evaluate(event: TriggerEvent): TriggeredEventRecord[] {
    const results: TriggeredEventRecord[] = [];
    const evalStart = performance.now();

    for (const [name, trigger] of this.triggers) {
      if (this.shouldTrigger(trigger, event)) {
        // Check cooldown
        const lastTrigger = this.cooldowns.get(name);
        const now = Date.now();

        if (lastTrigger && trigger.cooldown_seconds > 0) {
          const cooldownMs = trigger.cooldown_seconds * 1000;
          if (now - lastTrigger < cooldownMs) {
            continue; // Still in cooldown
          }
        }

        // Execute trigger
        const message = this.formatMessage(trigger.message || 'Trigger fired', event);
        this.executeAction(trigger, message, event);

        // Record trigger
        const triggerEvent: TriggeredEventRecord = {
          trigger_name: name,
          timestamp: Math.floor(now / 1000),
          message,
          action: trigger.action,
          file: event.file || null
        } as any;

        this.triggeredEvents.unshift(triggerEvent);
        if (this.triggeredEvents.length > 1000) {
          this.triggeredEvents.pop();
        }

        // Update cooldown and counts
        this.cooldowns.set(name, now);
        this.triggerCounts.set(name, (this.triggerCounts.get(name) || 0) + 1);

        results.push(triggerEvent);

        // Emit via EventBus
        const eventBusEvent: TriggerFiredEvent = {
          ruleName: name,
          message,
          severity: 'info',
          metadata: { action: trigger.action, trigger }
        };
        EventBus.emitTriggerFired(eventBusEvent);

        // Emit real-time trigger event via WebSocket
        if (this.io) {
          this.io.emit('trigger-fired', triggerEvent);
          this.io.emit('trigger-stats', this.getTriggerStats());
        }
      }
    }

    // Trim expired cooldown entries to prevent unbounded map growth
    this.trimExpiredCooldowns();

    internalMetrics.recordTriggerLatency(performance.now() - evalStart);
    return results;
  }

  /**
   * Remove cooldown entries that have expired past the max cooldown time
   */
  private trimExpiredCooldowns(): void {
    if (this.cooldowns.size <= this.triggers.size) return; // Nothing to trim

    const now = Date.now();
    let maxCooldownMs = 0;
    for (const trigger of this.triggers.values()) {
      maxCooldownMs = Math.max(maxCooldownMs, (trigger.cooldown_seconds || 60) * 1000);
    }

    for (const [name, lastFired] of this.cooldowns) {
      if (now - lastFired > maxCooldownMs) {
        this.cooldowns.delete(name);
      }
    }
  }

  /**
   * Check if event matches trigger conditions
   */
  private shouldTrigger(trigger: TriggerRule, event: TriggerEvent): boolean {
    // Check file pattern
    if (trigger.file && event.file) {
      if (!this.matchPattern(event.file, trigger.file)) {
        return false;
      }
    }

    // Check agent
    if (trigger.agent && event.agent) {
      if (event.agent.toLowerCase() !== trigger.agent.toLowerCase()) {
        return false;
      }
    }

    // Check event type
    if (trigger.event_type && event.event_type) {
      if (event.event_type.toLowerCase() !== trigger.event_type.toLowerCase()) {
        return false;
      }
    }

    // Numeric conditions: when a trigger declares one, the event MUST carry
    // the matching field, otherwise no match. The previous code treated
    // missing fields as auto-pass, so cpu_percent=">80" and lines_deleted=
    // ">100" rules fired on every plain file event (event had no cpu/duration
    // attached) — the recent-events feed showed "0%" / "0 lines deleted"
    // because the absent field was coerced to 0 in the message template.
    if (trigger.lines_changed) {
      if (
        event.lines_changed === undefined ||
        !this.checkNumericCondition(event.lines_changed, trigger.lines_changed)
      ) {
        return false;
      }
    }
    if (trigger.lines_deleted) {
      // lines_deleted reuses the lines_changed field on the event payload.
      if (
        event.lines_changed === undefined ||
        !this.checkNumericCondition(event.lines_changed, trigger.lines_deleted)
      ) {
        return false;
      }
    }
    if (trigger.duration_ms) {
      if (
        event.duration_ms === undefined ||
        !this.checkNumericCondition(event.duration_ms, trigger.duration_ms)
      ) {
        return false;
      }
    }
    if (trigger.cpu_percent) {
      if (
        event.cpu_percent === undefined ||
        !this.checkNumericCondition(event.cpu_percent, trigger.cpu_percent)
      ) {
        return false;
      }
    }
    if (trigger.memory_percent) {
      if (
        event.memory_percent === undefined ||
        !this.checkNumericCondition(event.memory_percent, trigger.memory_percent)
      ) {
        return false;
      }
    }

    return true;
  }

  /**
   * Match filename against glob pattern
   */
  private matchPattern(filename: string, pattern: string): boolean {
    // Simple glob pattern matching
    const regex = pattern.replace(/\./g, '\\.').replace(/\*/g, '.*').replace(/\?/g, '.');
    return new RegExp(`^${regex}$`).test(filename);
  }

  /**
   * Check numeric condition (>, <, >=, <=, ==)
   */
  private checkNumericCondition(value: number, condition: string): boolean {
    const match = condition.match(/^([><=]+)(\d+\.?\d*)$/);
    if (!match) return false;

    const [, operator, threshold] = match;
    const thresholdNum = parseFloat(threshold);

    switch (operator) {
      case '>':
        return value > thresholdNum;
      case '<':
        return value < thresholdNum;
      case '>=':
        return value >= thresholdNum;
      case '<=':
        return value <= thresholdNum;
      case '==':
        return value === thresholdNum;
      default:
        return false;
    }
  }

  /**
   * Format message template with event values
   */
  private formatMessage(template: string, event: TriggerEvent): string {
    return (
      template
        .replace(/\{file\}/g, String(event.file || ''))
        .replace(/\{agent\}/g, String(event.agent || ''))
        .replace(/\{event_type\}/g, String(event.event_type || ''))
        .replace(/\{lines_changed\}/g, String(event.lines_changed || 0))
        // File events carry no separate deletion count — `lines_changed` (the
        // diff's line count) is all we have, and the `lines_deleted` trigger
        // condition reuses it (see matches()). So the {lines_deleted} placeholder
        // resolves to the same field; otherwise it read a field that never exists
        // and always rendered 0 ("…0 lines deleted").
        .replace(/\{lines_deleted\}/g, String(event.lines_changed || 0))
        .replace(/\{duration_ms\}/g, String(event.duration_ms || 0))
        .replace(/\{cpu_percent\}/g, String(event.cpu_percent || 0))
        .replace(/\{memory_percent\}/g, String(event.memory_percent || 0))
    );
  }

  /**
   * Execute trigger action (log, notify, command)
   */
  private async executeAction(
    trigger: TriggerRule,
    message: string,
    event: TriggerEvent
  ): Promise<void> {
    try {
      switch (trigger.action) {
        case 'notify':
          await this.sendNotification(message);
          break;
        case 'log':
          this.logToFile(trigger.name, message);
          break;
        case 'command':
          if (trigger.command) {
            await this.executeCommand(trigger.command, event);
          }
          break;
      }
    } catch (error) {
      logger.error(`❌ Failed to execute trigger action: ${error}`);
    }
  }

  /**
   * Sanitize user input to prevent command injection
   */
  private sanitizeShellArg(arg: string | number): string {
    let str = typeof arg !== 'string' ? String(arg) : arg;
    return str
      .replace(/[;&|`$(){}[\]<>\\'"]/g, '')
      .replace(/\n/g, ' ')
      .replace(/\r/g, '')
      .trim()
      .slice(0, 500);
  }

  /**
   * Format message safely by sanitizing all replaced values
   */
  private formatMessageSafe(template: string, event: TriggerEvent): string {
    return template
      .replace(/\{file\}/g, event.file ? this.sanitizeShellArg(event.file) : '')
      .replace(/\{agent\}/g, event.agent ? this.sanitizeShellArg(event.agent) : '')
      .replace(/\{event_type\}/g, event.event_type ? this.sanitizeShellArg(event.event_type) : '')
      .replace(
        /\{lines_changed\}/g,
        event.lines_changed !== undefined ? this.sanitizeShellArg(event.lines_changed) : ''
      )
      .replace(
        /\{duration_ms\}/g,
        event.duration_ms !== undefined ? this.sanitizeShellArg(event.duration_ms) : ''
      )
      .replace(
        /\{cpu_percent\}/g,
        event.cpu_percent !== undefined ? this.sanitizeShellArg(event.cpu_percent) : ''
      )
      .replace(
        /\{memory_percent\}/g,
        event.memory_percent !== undefined ? this.sanitizeShellArg(event.memory_percent) : ''
      );
  }

  /**
   * Send platform-specific notification
   */
  private async sendNotification(message: string): Promise<void> {
    try {
      const sanitized = this.sanitizeShellArg(message);
      if (process.platform === 'linux') {
        await execFileAsync('notify-send', ['Raven Trigger', sanitized], {
          timeout: 5000,
          shell: false
        });
      } else if (process.platform === 'darwin') {
        await execFileAsync(
          'osascript',
          ['-e', `display notification "${sanitized}" with title "Raven Trigger"`],
          { timeout: 5000, shell: false }
        );
      } else {
        logger.info(`📢 Notification: ${message}`);
      }
    } catch (error) {
      logger.error(`❌ Failed to send notification: ${error}`);
    }
  }

  /**
   * Log trigger to file
   */
  private logToFile(triggerName: string, message: string): void {
    try {
      const logFile = path.join(this.configPath, 'triggers.log');
      const timestamp = new Date().toISOString();
      const logLine = `[${timestamp}] ${triggerName} - ${message}\n`;

      fs.appendFileSync(logFile, logLine);
      logger.info(`📝 Logged: ${message}`);
    } catch (error) {
      logger.error(`❌ Failed to log to file: ${error}`);
    }
  }

  private static readonly ALLOWED_COMMANDS = ['echo', 'notify-send', 'osascript', 'logger', 'wall'];

  /**
   * Execute shell command with whitelist and argument sanitization
   */
  private async executeCommand(command: string, event: TriggerEvent): Promise<void> {
    try {
      const formattedCommand = this.formatMessageSafe(command, event);
      const parts = formattedCommand.trim().split(/\s+/);
      const baseCommand = parts[0];
      const args = parts.slice(1);

      const isAllowed = TriggerEngine.ALLOWED_COMMANDS.some(
        allowed => baseCommand === allowed || baseCommand.endsWith(`/${allowed}`)
      );

      if (!isAllowed) {
        logger.error(`❌ Command execution blocked: '${baseCommand}' is not in whitelist`);
        logger.error(`   Allowed commands: ${TriggerEngine.ALLOWED_COMMANDS.join(', ')}`);
        return;
      }

      const sanitizedArgs = args.map(arg => this.sanitizeShellArg(arg));

      await execFileAsync(baseCommand, sanitizedArgs, {
        timeout: 5000,
        shell: false
      });

      logger.info(`⚙️  Executed command: ${baseCommand} ${sanitizedArgs.join(' ')}`);
    } catch (error) {
      logger.error(
        `❌ Failed to execute command: ${error instanceof Error ? error.message : error}`
      );
    }
  }

  // ==================== API Methods ====================

  /**
   * Get all trigger configurations
   */
  getTriggersConfig(): TriggerRule[] {
    return Array.from(this.triggers.values());
  }

  /**
   * Get recent triggered events
   */
  getTriggeredEvents(limit: number = 100): TriggeredEventRecord[] {
    return this.triggeredEvents.slice(0, limit);
  }

  /**
   * Get trigger statistics
   */
  getTriggerStats(): TriggerStats {
    return {
      total_triggers: this.triggeredEvents.length,
      active_triggers: this.triggers.size,
      trigger_counts: Object.fromEntries(this.triggerCounts)
    };
  }

  /**
   * Reload configuration from disk
   */
  reloadConfig(): string {
    this.loadConfig();
    return `Reloaded ${this.triggers.size} trigger rules`;
  }

  /**
   * Clear all cooldowns
   */
  clearCooldowns(): string {
    this.cooldowns.clear();
    return 'All trigger cooldowns cleared';
  }

  /**
   * Clear all trigger counts
   */
  clearStats(): string {
    this.triggerCounts.clear();
    return 'All trigger counts cleared';
  }
}
